content/blog/2024/20240422_pwm_rpi5.md

```diff
@@ -0,0 +1,196 @@
+
+
+```text
+[all]
+dtoverlay=pwm-2chan,pin=12,func=4,pin2=13,func2=4
+```
+
+For example, see screenshot on how to edit `config.sys` to set to a single PWM GPIO18:
+
+![](/assets/blogs/pwm/pwm-config-txt.jpg)
+
+### Reboot
+
+* Close your open applications.
+* Enter: `sudo shutdown -r`. This will terminate all remaining applications and reboot.
+
+### Modify the application
+
+* Your system is now ready to use the `linuxfs-pwm` provider.
+* Previously the application's `PwmConfigBuilder address was the GPIO number, but now you need to use the channel number as the address parameter.
+
+```java
+package com.pi4j.test.devices.pwmTest;
+
+import com.pi4j.Pi4J;
+import com.pi4j.context.Context;
+import com.pi4j.io.pwm.Pwm;
+import com.pi4j.io.pwm.PwmType;
+import com.pi4j.util.Console;
+
+import java.util.Scanner;
+
+public class UserTest {
+    private static Pwm pwm = null;
+    private static Context pi4j;
+
+    public static void main(String[] args) throws Exception {
+        pi4j = Pi4J.newAutoContext(); 
+
+        var console = new Console();
+        System.out.println("----------------------------------------------------------");
+        System.out.println("PI4J PROVIDERS");
+        System.out.println("----------------------------------------------------------");
+        pi4j.providers().describe().print(System.out);
+        System.out.println("----------------------------------------------------------");
+
+        int address = 2;
+        //  initPiGpio();
+        for (int i = 0; i < args.length; i++) {
+            String o = args[i];
+            if (o.contentEquals("-a")) { // pin
+                String a = args[i + 1];
+                address = Short.parseShort(a.substring(0), 10);
+                i++;
+            } else {
+                console.println("  No valid Parm, -a address ");
+                System.exit(42);
+            }
+        }
+        initGPIO(address);
+
+        System.out.println("linuxfs pin after creation at duty_cycle 50  ");
+        waitForInput(console);
+
+        pwm.on(50, 2);
+        System.out.println("linuxfs pin after pin.on  freq 2    ");
+
+        waitForInput(console);
+
+        pwm.off();
+        pwm.on(50, 10);
+        System.out.println("linuxfs pin after pin.on  freq 10  ");
+        waitForInput(console);
+
+        pwm.off();
+        pwm.on(50, 50);
+        System.out.println("linuxfs pin after pin.on  freq 50   ");
+
+        waitForInput(console);
+
+        pwm.off();
+        pwm.on(50, 1);
+        System.out.println("linuxfs pin after pin.on  freq 1");
+        waitForInput(console);
+
+
+        System.out.println("linuxfs pin call pwm.shutdown ");
+        waitForInput(console);
+        
+        pwm.shutdown(pi4j);
+
+        System.out.println("linuxfs pi4j  call pi4j.shutdown ");
+        waitForInput(console);
+
+        pi4j.shutdown();
+        while (true) {
```

Comment:
This is a active spin, which would use a lot of CPU resources. 

We can simply to a Thread.sleep(LONG_TIME) or a another waitForInput.